src/middlewared/middlewared/plugins/app_registry/crud.py

```diff
@@ -0,0 +1,87 @@
+import middlewared.sqlalchemy as sa
+from middlewared.api import api_method
+from middlewared.api.current import (
+    AppRegistryEntry, AppRegistryCreateArgs, AppRegistryCreateResult, AppRegistryUpdateArgs,
+    AppRegistryUpdateResult, AppRegistryDeleteArgs, AppRegistryDeleteResult,
+)
+from middlewared.service import CRUDService, private, ValidationErrors
+
+from .validate_registry import validate_registry_credentials
+
+
+class AppRegistryModel(sa.Model):
+    __tablename__ = 'app_registry'
+
+    id = sa.Column(sa.Integer(), primary_key=True)
+    name = sa.Column(sa.String(255), nullable=False)
+    description = sa.Column(sa.String(512), nullable=True, default=None)
+    username = sa.Column(sa.EncryptedText(), nullable=False)
+    password = sa.Column(sa.EncryptedText(), nullable=False)
+    uri = sa.Column(sa.String(512), nullable=False, unique=True)
+
+
+class AppRegistryService(CRUDService):
+
+    class Config:
+        namespace = 'app.registry'
+        datastore = 'app.registry'
+        cli_namespace = 'app.registry'
+        entry = AppRegistryEntry
+        # FIXME: Add roles support
```

Comment:
This needs to be added _before_ merge. I'm getting ready to have middleware break on public CRUD/ config services without role_prefix defined.

Reply:
NOTE: you can have multiple services with same role_prefix (e.g. "APPS").